src/transformers/modeling_tf_pytorch_utils.py

```diff
@@ -284,14 +257,151 @@ def load_pytorch_weights_in_tf2_model(tf_model, pt_state_dict, tf_inputs=None, a
+        # -> PT: (num_out_channel, num_in_channel, kernel[0], kernel[1])
+        array = np.transpose(array, axes=(3, 2, 0, 1))
+    elif transpose is TransposeType.CONV1D:
+        # Conv1D weight:
+        #    TF: (kernel, num_in_channel, num_out_channel)
+        # -> PT: (num_out_channel, num_in_channel, kernel)
+        array = np.transpose(array, axes=(2, 1, 0))
+    elif transpose is TransposeType.SIMPLE:
+        array = np.transpose(array)
+
+    if len(ptr.shape) < len(array.shape):
+        array = array.squeeze()
+    elif len(ptr.shape) > len(ptr.shape):
+        array = np.expand_dims(array, axis=0)
+    if list(ptr.shape) != list(array.shape):
+        squeezed_tensor_shape = [dim for dim in ptr.shape if dim != 1]
+        squeezed_array_shape = [dim for dim in array.shape if dim != 1]
+        # Making sure we don't do wrong reshape
+        if squeezed_tensor_shape == squeezed_array_shape:
+            try:
+                array = array.reshape(ptr.shape)
+            except AssertionError as e:
+                raise e
+    # logger.warning(f"Initialize PyTorch weight {pt_weight_name}")
+    # Make sure we have a proper numpy array
+    if np.isscalar(array):
+        array = np.array(array)
+    return array
+
+
+def apply_reverse(array, transpose, symbolic_weight):
+    """
+    Given a numpy array, a transpose enum, and the target tensor shape Attempt to modify the array into something
+    fitting the target tensor The direction here is PT -> TF
+    """
+    if transpose is TransposeType.CONV2D:
+        # Conv2D weight:
+        #    PT: (num_out_channel, num_in_channel, kernel[0], kernel[1])
+        # -> TF: (kernel[0], kernel[1], num_in_channel, num_out_channel)
+        array = np.transpose(array, axes=(2, 3, 1, 0))
+    elif transpose is TransposeType.CONV1D:
+        # Conv1D weight:
+        #    PT: (num_out_channel, num_in_channel, kernel)
+        # -> TF: (kernel, num_in_channel, num_out_channel)
+        array = np.transpose(array, axes=(2, 1, 0))
+    elif transpose is TransposeType.SIMPLE:
+        array = np.transpose(array)
+
+    if len(symbolic_weight.shape) < len(array.shape):
+        array = np.squeeze(array)
+    elif len(symbolic_weight.shape) > len(array.shape):
+        array = np.expand_dims(array, axis=0)
+
+    if list(symbolic_weight.shape) != list(array.shape):
+        squeezed_tensor_shape = [dim for dim in symbolic_weight.shape if dim != 1]
+        squeezed_array_shape = [dim for dim in array.shape if dim != 1]
+        if squeezed_array_shape == squeezed_tensor_shape:
+            try:
+                array = np.reshape(array, symbolic_weight.shape)
+            except AssertionError as e:
+                e.args += (symbolic_weight.shape, array.shape)
+                raise e
+
+    try:
+        assert list(symbolic_weight.shape) == list(array.shape)
+    except AssertionError as e:
+        e.args += (symbolic_weight.shape, array.shape)
+        raise e
+    return array
```

Comment:
I think it is still useful (in this PR, not the follow up), since if dimensions don't match in the end we will crash here instead of later when attempting to write the data to the weight in the model.

In the follow up, `apply_transpose` don't have the reference to the tensor anymore so this check cannot exist (but operation is easily invertible)